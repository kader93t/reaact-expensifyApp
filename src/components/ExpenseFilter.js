import React from 'react'
import { connect } from 'react-redux';
import {setEndDate, setStartDate, setTextFilter} from "./../actions/filter"
import { setSortByDate, setSortByAmount } from "./../actions/filter"
import { DateRangePicker } from "react-dates"


export class ExpensesFilters extends React.Component{
    state = {
        focusedInput: undefined
    }

    setSortByDate = () => {
        this.props.setSortByDate();
    }

    setSortByAmount = () => {
        this.props.setSortByAmount();
    }

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    setTextFilter = (e) => {
        this.props.setTextFilter(e.target.value);
    }

    onSortChange = (e) => {      
        switch (e.target.value) {
            case "date": {
                this.setSortByDate();
                break;
            }
            case "amount":{
                this.setSortByAmount();
                break;
            }
        }
    }
    
    render() {
        
        return  ( <div>        
            
            <input type="text" placeholder="Text filter" value={this.props.filter.text} onChange={this.setTextFilter}/>
            
            <select
                value={this.props.filter.sortBy}
                onChange={this.onSortChange}>
                <option value="amount">Amount</option>
                <option value="date">Date</option>
            
            </select>
            
            <DateRangePicker
                startDate={this.props.filter.startDate}
                startDateId="startDate"
                endDate={this.props.filter.endDate}
                endDateId="endDate"
                onDatesChange={
                    this.onDatesChange
                }
                focusedInput={this.state.focusedInput} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                onFocusChange={focusedInput => this.setState({ focusedInput })} // PropTypes.func.isRequired,
                isOutsideRange={() => false}
                showClearDates={true}
                numberOfMonths = {1}
        />
            </div>)
            }
}

const mapStateToProps = (state) => {
    return {
        filter: state.filter    
    }
};

const mapDispatchToProps = (dispatch) => ({
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setSortByAmount: () => dispatch(setSortByAmount()),
    setSortByDate: () => dispatch(setSortByDate()),
    setTextFilter: (text) => dispatch(setTextFilter(text))
});
export default connect(mapStateToProps, mapDispatchToProps)(ExpensesFilters);

