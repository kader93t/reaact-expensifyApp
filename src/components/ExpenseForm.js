import React, { Component } from "react";
import { SingleDatePicker } from "react-dates";

import moment from "moment";

export default class ExpenseForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? props.expense.amount.toString() : "",
      createAt: props.expense ? moment(props.expense.createAt) : moment(),
      calendarFocused: false,
      error: "",
    };
  }

  onChangeDescription = (e) => {
    const description = e.target.value;
    this.setState((oldState) => ({
      ...oldState,
      description,
    }));
  };

  onChangeNote = (e) => {
    const note = e.target.value;
    this.setState((oldState) => ({
      ...oldState,
      note,
    }));
  };

  onChangeAmount = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,3})?$/)) {
      this.setState((oldState) => ({
        ...oldState,
        amount,
      }));
    }
  };

  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState((oldState) => ({
        error: "One of the important field is empty",
      }));
    } else {
      this.setState((oldState) => ({
        error: "",
      }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount),
        createAt: this.state.createAt.valueOf(),
        note: this.state.note,
      });
    }
  };

  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          className="text-input"
          type="text"
          placeholder="Description"
          autoFocus
          value={this.state.description}
          onChange={this.onChangeDescription}
        />
        <input
          className="text-input"
          type="text"
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.onChangeAmount}
        />

        <textarea
          className="input-area"
          placeholder="Add a note for your expense (optional)"
          value={this.state.note}
          onChange={this.onChangeNote}
        ></textarea>

        <SingleDatePicker
          date={this.state.createAt}
          onDateChange={(createAt) => {
            if (createAt) this.setState(() => ({ createAt }));
          }}
          focused={this.state.calendarFocused}
          onFocusChange={({ focused: calendarFocused }) => {
            this.setState(() => ({ calendarFocused }));
          }}
          id="my_id"
          isOutsideRange={() => false}
          numberOfMonths={1}
        />
        <div >
          <button className='button'> Add Expnese </button>
        </div>
      </form>
    );
  }
}
