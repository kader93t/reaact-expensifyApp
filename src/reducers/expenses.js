// Expenses Reducer

const expenseDefaultState = [];
export default (state = expenseDefaultState, action) => {
    
    switch (action.type) {
        case "ADD_EXPENSE":
            return [       
                ...state, action.expenses
            ]
        case "REMOVE_EXPENSE":
            return state.filter((expense) => {
                return expense.id !== action.id;
            })
        case "UPDATE_EXPENSE":
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    }
                } else {
                    return expense;
                }
            });
            ;
        default:
            return state;
    }
}


