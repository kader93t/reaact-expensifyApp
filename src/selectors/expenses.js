
export default (expenses, { startDate, endDate, text, sortBy }) => {
    const visibleExpenses = expenses.filter((expense) => {
        const matchStartDate = !startDate || startDate.isSameOrBefore(expense.createAt, "day");
        const matchEndDate = !endDate || endDate.isSameOrAfter(expense.createAt, "day");
        const matchText = !text || expense.description.toLowerCase().includes(text.toLowerCase());
        return matchEndDate && matchStartDate && matchText;
    });
    return visibleExpenses.sort((a, b) => {
        if (sortBy === "date")
            return a.createAt < b.createAt ? 1 : -1;
        else if (sortBy === "amount")
            return a.amount < b.amount ? 1 : -1;
    });
};
