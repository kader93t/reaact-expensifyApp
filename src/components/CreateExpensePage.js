import React, { Component } from "react";
import { connect } from "react-redux";
import { startAddExpense } from "./../actions/expneses.js";
import ExpenseForm from "./ExpenseForm";

export class CreateExpensePage extends Component {
  onSubmit = (expense) => {
    this.props.onSubmit(expense);
    this.props.history.push("/");
  };

  render() {
    return (
      <div>
        <div className="summary">
          <div className="content-container">
            <h1 className="summary-page__title">Add expense </h1>
          </div>
        </div>
        <div className= "content-container">
          <ExpenseForm onSubmit={this.onSubmit} />
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  onSubmit: (expense) => dispatch(startAddExpense(expense)),
});

export default connect(undefined, mapDispatchToProps)(CreateExpensePage);
