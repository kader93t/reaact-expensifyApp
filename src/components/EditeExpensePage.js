import React from "react";
import { connect } from "react-redux";
import { startEditeExpense, startRemoveExpense } from "../actions/expneses";
import ExpenseForm from "./ExpenseForm";

export class EditeExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.modifieExpense(this.props.expense.id, expense);
    this.props.history.push("/");
  };

  onButtonClick = (e) => {
    this.props.onButtonClick(this.props.expense.id);
    this.props.history.push("/");
  };

  render() {
    return (
      <div>
        <div className= "summary">
          <div className=" content-container">
            <h1 className="summary-page__title"> Edit expense</h1>
          </div>
        </div>

        <div className="content-container">
          <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
          <button className="button__second" onClick={this.onButtonClick}>
            {" "}
            Remove{" "}
          </button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(
    (expense) => expense.id === props.match.params.id
  ),
});

const mapDispatchToProps = (dispatch) => ({
  modifieExpense: (id, expense) => dispatch(startEditeExpense(id, expense)),
  onButtonClick: (id) => dispatch(startRemoveExpense(id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditeExpensePage);
