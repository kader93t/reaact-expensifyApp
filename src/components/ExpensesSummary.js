import React from "react";
import ReactDOM from "react-dom";
import selectExpenses from "./../selectors/expenses";
import { connect } from "react-redux";
import { expenseSummary } from "./../selectors/expense-total";
import { Link } from "react-router-dom";
import numeral from "numeral";
export const ExpensesSummary = (props) => {
  const [numberOfExpense, total] = expenseSummary(props.expenses);
  const expenseWord = numberOfExpense == 1 ? "expense" : "expenses";
  return (
    <div className="summary">
      <div className="content-container">
        <p className="summary-page__title">
          Viewing : <span> {numberOfExpense} </span> {expenseWord} , The total :{" "}
          <span>{numeral(total).format("$0,0.00")}</span>
        </p>

        <div className="summary-page__actions">
          <Link className="button" to="/create">
            {" "}
            Add expense
          </Link>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = (state) => ({
  expenses: selectExpenses(state.expenses, state.filter),
});

export default connect(mapStateToProps)(ExpensesSummary);
