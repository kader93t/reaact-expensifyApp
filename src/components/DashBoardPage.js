import React from "react"
import ExpensesList from './ExpnesesList'
import ExpenseFilter from "./ExpenseFilter"
import  ExpensesSummary  from "./ExpensesSummary";

const ExpenseDashBoardPage = () => (
    <div>
         <ExpensesSummary />
        <ExpenseFilter />
        <ExpensesList />
    </div>
);

export default ExpenseDashBoardPage;