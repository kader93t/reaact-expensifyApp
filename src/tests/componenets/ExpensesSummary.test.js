import React from 'react'
import { shallow } from "enzyme";
import { expenses } from "../fixture/expenses";
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('should render the total summary ', () => {
    const wrapper = shallow(<ExpensesSummary expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
});

test('should rnder the sammary with one expense', () => {
    const wrapper = shallow(<ExpensesSummary expenses={[expenses[0]]} />);
    expect(wrapper).toMatchSnapshot();
});


