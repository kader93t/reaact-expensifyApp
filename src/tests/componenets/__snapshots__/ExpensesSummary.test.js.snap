// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the total summary  1`] = `
<div
  className="summary"
>
  <div
    className="content-container"
  >
    <p
      className="summary-page__title"
    >
      Viewing : 
      <span>
         
        3
         
      </span>
       
      expenses
       , The total :
       
      <span>
        $6.36
      </span>
    </p>
    <div
      className="summary-page__actions"
    >
      <Link
        className="button"
        to="/create"
      >
         
        Add expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should rnder the sammary with one expense 1`] = `
<div
  className="summary"
>
  <div
    className="content-container"
  >
    <p
      className="summary-page__title"
    >
      Viewing : 
      <span>
         
        1
         
      </span>
       
      expense
       , The total :
       
      <span>
        $2.12
      </span>
    </p>
    <div
      className="summary-page__actions"
    >
      <Link
        className="button"
        to="/create"
      >
         
        Add expense
      </Link>
    </div>
  </div>
</div>
`;
